basesandbox/src/ipc/unix_socket.rs

```diff
@@ -0,0 +1,381 @@
+use super::*;
+use crossbeam::channel::{self, Receiver, Sender};
+use mio::net::{UnixListener, UnixStream};
+use mio::{Events, Interest, Poll, Token};
+use parking_lot::Mutex;
+use std::io::{Read, Write};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::thread;
+
+const POLL_TOKEN: Token = Token(0);
+
+// TODO: Separate 8 bytes size prefix, which is for packet framing, out of pure streaming implementation
+// and let it stay as just one of instance of the communication protocol.
+
+fn send_routine(
+    queue: Receiver<Vec<u8>>,
+    write_signal: Receiver<Result<(), ()>>,
+    socket: Arc<Mutex<UnixStream>>,
+) -> Result<(), String> {
+    #[derive(Debug)]
+    enum Error {
+        ExpectedTermination,
+        UnexpectedError(String),
+    }
+
+    let send_helper = |buf: &[u8]| {
+        let mut sent = 0;
+        while sent < buf.len() {
+            // NOTE: Never replace r directly into match.
+            // If then, the mutex will be locked during the whole match statement!
+            let r = socket.lock().write(&buf[sent..]);
```

Comment:
This is not a change request.

I think the code below gives enough information that the programmer who wrote the code wants to reduce the region of the lock. So it will not need the comment.

However, I don't know whether other people agree with me or not. Please change it if you agree with me. If you don't agree with me you don't need to change the code.

```suggestion
            let r = {
              let lock_guard = socket.lock();
              lock_guard.write(&buf[sent..]);
            }
```

Reply:
@majecty If there was no `match`, then I wouldn't left that note. It is more like Rust tip for the match evaluation. If I just put your code removing the code, someone still might think "I know that the lock is locked for only the `r` evaluation, but why not just put in the match? It will be same".

Reply:
OK. Let's left it.